File: cp-algorithms/dfs/cpp/C_kefaAndPark.cpp
#include <iostream>
#include <vector>

const int maxn = 100005;
int N, M, ans;
std::vector<int> e[maxn];
bool cat[maxn];

void dfs(int u, int fa, int cnt) {
    if (cnt > M)
        return;
    bool is_leaf = 1;
    for (auto v : e[u]) {
        if (v == fa) 
            continue;
        is_leaf = 0;
        dfs(v, u, cat[v]?cnt+1:0);
    }
    ans += is_leaf;
}

int main() {
    std::ios::sync_with_stdio(0); std::cin.tie(0);
    std::cin >> N >> M;
    for (int i = 1; i <= N; i++) 
        std::cin >> cat[i];
    for (int i = 1; i < N; i++) {
        int u, v;
        std::cin >> u >> v;
        e[u].push_back(v);
        e[v].push_back(u);
    }
    dfs(1, 0, cat[1]);
    std::cout << ans << '\n';
    return 0;
}

File: cp-algorithms/dfs/python/C_kefaAndPark.py
def main():
    N, M = map(int, input().split())
    g = [[] for _ in range(N+1)]
    cat = [0] + list(map(int, input().split()))
    for i in range(1, N):
        u, v = map(int, input().split())
        g[u].append(v)
        g[v].append(u)
    ans = 0

    def dfs(u: int, fa: int, cnt: int):
        nonlocal ans
        if cnt > M:
            return 
        is_leaf = True
        for v in g[u]:
            if v == fa:
                continue
            is_leaf = False
            c = cnt + 1 if cat[v] else 0
            dfs(v, u, c)
        ans += is_leaf

    dfs(1, 0, cat[1])
    print(ans)
        
main()
    


